vowpalwabbit/shared_feature_merger.cc

```diff
@@ -0,0 +1,106 @@
+#include "shared_feature_merger.h"
+#include "cb.h"
+#include "example.h"
+#include "label_dictionary.h"
+#include "learner.h"
+#include "options.h"
+#include "parse_args.h"
+#include "vw.h"
+
+#include <iterator>
+
+namespace VW
+{
+namespace shared_feature_merger
+{
+static const std::vector<std::string> option_strings = {
+    "csoaa_ldf", "wap_ldf", "cb_adf", "explore_eval", "cbify_ldf", "cb_explore_adf", "warm_cb"};
+
+bool use_reduction(config::options_i& options)
+{
+  for (const auto& opt : option_strings)
+  {
+    if (options.was_supplied(opt))
+      return true;
+  }
+  return false;
+}
+
+struct sfm_data
+{
+  LEARNER::multi_learner* base;
+};
+
+template <bool is_learn>
+void predict_or_learn(sfm_data& data, LEARNER::multi_learner& base, multi_ex& ec_seq)
+{
+  if (ec_seq.size() == 0)
+    THROW("cb_adf: At least one action must be provided for an example to be valid.");
+
+  multi_ex::value_type shared_example = nullptr;
+
+  const bool has_example_header = CB::ec_is_example_header(*ec_seq[0]);
+  if (has_example_header)
+  {
+    shared_example = ec_seq[0];
+    ec_seq.erase(ec_seq.begin());
+    // merge sequences
+    for (auto & example : ec_seq) LabelDict::add_example_namespaces_from_example(*example, *shared_example);
+    ec_seq[0]->pred = shared_example->pred;
+  }
+  if (ec_seq.size() == 0)
+    return;
+  if (is_learn)
+    base.learn(ec_seq);
+  else
+    base.predict(ec_seq);
+
+  if (has_example_header)
+  {
+    for (auto& example : ec_seq) LabelDict::del_example_namespaces_from_example(*example, *shared_example);
+    shared_example->pred = ec_seq[0]->pred;
```

Comment:
If you add example->num_features += shared_example->num_features at this line, you'll have a better accounting of the number of features (although it will not account for quadratics).

Reply:
Since the examples can be passed through the reductions more than once, and since we're passing the examples by pointers here, I'm not sure its safe to modify the num_features directly without having a mechanism to undo the change elsewhere.

If I attempt to do the same operation in finish_multiline_example(), several other (seemingly unrelated) tests fail with bad feature counts.
As it is right now, the `current features` are the same as before this PR. The only thing that changes is the `total feature number` on several tests

Reply:
In the long term, I think we want the update rule itself to keep track of the number of features, in which case multiple uses of the same example results in incrementing.  

I agree this is not essential, but I don't know a good reason to not do it.

Reply:
In that case, the feature count increment should occur in finish_example, not predict_or_learn.

Incrementing the count here will change the number of features seen by upstream reductions. If an example is used in both predict and learn, we'd be incrementing these features multiple times as well.

Reply:
That seems fine.